[3 - 4단계 방탈출 예약 대기] 띠용(주용은) 미션 제출합니다.

jye326 opened the pull request:
<!-- 

## 코드 리뷰 팁

- 코드와 관련된 질문이 있다면, PR 본문에 적기 보다는 해당 코드를 선택하고 코멘트를 남겨주세요.
  - [참고: Adding comments to a pull request](https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/reviewing-changes-in-pull-requests/commenting-on-a-pull-request#adding-comments-to-a-pull-request)

-->
## 체크 리스트
- [x] 미션의 필수 요구사항을 모두 구현했나요?
- [x] Gradle `test`를 실행했을 때, 모든 테스트가 정상적으로 통과했나요?
- [x] 애플리케이션이 정상적으로 실행되나요?
- [ ] [프롤로그](https://prolog.techcourse.co.kr)에 셀프 체크를 작성했나요?
  <!-- 작성한 셀프 체크의 링크를 남겨주세요. -->

## 객체지향 생활체조 요구사항을 얼마나 잘 충족했다고 생각하시나요?
### 1~5점 중에서 선택해주세요.

- [ ] 1 (전혀 충족하지 못함)
- [ ] 2
- [x] 3 (보통)
- [ ] 4
- [ ] 5 (완벽하게 충족)

### 선택한 점수의 이유를 적어주세요.
<!-- 이유 작성 -->
객체 지향을 신경 쓰지 않았습니다.

## 베이스 코드 선택 체크
- [ ] 이전 미션의 내 코드에서 시작 
- [x] 이전 미션의 페어의 코드에서 시작

## 어떤 부분에 집중하여 리뷰해야 할까요?
<!-- 리뷰어가 효과적으로 피드백할 수 있도록 중점적으로 리뷰 받고 싶은 내용을 *요약* 형태로 작성해 주세요.
리뷰해야 할 포인트를 요약해 강조해 주시면, 리뷰어가 코드 전체를 이 부분에 집중해 리뷰할 수 있습니다.
반면 특정 코드부분에 대한 피드백이 필요하다면, 이 곳에 적기 보다 해당 코드를 선택하고 코멘트를 남기는 것을 권장합니다. -->

1. LMS의 구현 힌트를 보고 처음에는 예약 대기를 Waiting 클래스를 만들어 관리하도록 구현하다가, 고민 끝에 Reservation 객체의 상태값(ENUM)으로 관리하도록 구현했습니다. 방탈출예약관리라는 지금의 도메인에 대해서 Waiting 클래스 분리는 필요 없다고 생각합니다. README에 관련한 고민을 의식의 흐름 기법으로 적어 놓았습니다.  관련해서 의견을 듣고 싶습니다!

2. 많은 크루들이 이번 미션 중에 OSIV, 트랜잭션, 프록시 객체 등의 용어를 공부하며 뜨겁게 의견을 교류하는데 저는 이번 미션을 진행하며 이러한 개념을 체감할 수 있는 부분을 찾지 못했습니다. 미션에서 학습 주제로 제시했던 영속성 컨텍스트 또한 마찬가지입니다. 저도 미션을 통해 관련 개념에 대한 이해가 필요한 문제를 겪으며 공부해보고 싶습니다! 혹시 제 구현 방식에 문제가 있었는지, 아니면 현재 코드에서도 관련된 개념이 사용되고 있는 부분이 있는데 놓치고 있는 건지 궁금합니다!

3. 이번 미션에서 생성형 AI의 도움을 (많이)받아 JPQL을 활용했습니다.

### 📌 코드리뷰 질문 그라운드룰 (지우지 마세요)
✅ 많은 사람과 공유하고 토론해보는 것이 더 좋은 질문에는, 리뷰어가 💬말풍선 이모지를 남겨 토론을 추천할 수 있어요.  
✅ 토론을 추천받은 질문은 `#be-7기-잡담-질답`슬랙채널이나 캠퍼스에서 자유롭게 토론을 이어가보세요!  
✅ 코드에 대한 맥락 없이 리뷰어의 주관적인 의견을 묻거나, 단순히 현업에서는 어떻게 하는지를 묻는 질문은 코드 리뷰 본문에는 어울리지 않아요.  
코드 리뷰는 변경된 코드의 목적, 의도, 품질을 중심으로 논의하는 자리예요.  
✅ 코드와 관련된 질문이 있다면 PR 본문에 적기보다는, 해당 코드를 선택해 코멘트를 남겨주세요.

jye326 committed:
initial commit

- dompoo code 복사!

jye326 committed:
feat: VO를 JPA 칼럼으로 매핑

jye326 committed:
feat: 엔티티를 JPA 테이블로 매핑

jye326 committed:
feat: JPA 리포지토리 구현

jye326 committed:
feat: Reservation 엔티티를 테이블에 매핑, 리포지토리 구현

jye326 committed:
fix: 예약 일자 객체 검증 시점 변경

- 최초 생성 시에만 검증

jye326 committed:
feat: 값 객체 Embeddable 어노테이션 추가

jye326 committed:
test: infrastructure 테스트 jpa 전환

jye326 committed:
fix : 쿼리수정, 어노테이션 추가

jye326 committed:
delete: 미사용 resources 삭제

jye326 committed:
fix: 예약된 시간 응답이 논리적으로 반대였던 오류 해결

jye326 committed:
feat: 내 예약 기능의 html, js 추가

jye326 committed:
feat: 내 예약 api 작성

dompoo committed:
style: 코드 스타일 통일

jye326 committed:
refactor(repository) : 패키지 분리

- JDBC, JPA 리포지토리 패키지 분리

jye326 committed:
fix(entity) : 기본 생성자 접근 지정 변경 -> protected

jye326 committed:
feat(entity) : 테이블 명 명시

jye326 committed:
feat(entity) : UserRole 저장 형식 변경 (ORDINAL-> STRING)

jye326 committed:
refactor: Jdbc 레포지토리 코드 제거 및 Jpa 레포지토리로 통합

- Jdbc 기반의 Reservation, ReservationTime, Theme, User 레포지토리 코드를 삭제하고 기존 Jpa 기반 레포지토리로 통합.
- 관련 테스트 코드 및 인프라 패키지 구조에서 불필요한 "Jpa" 네이밍 제거로 간결화 작업 수행.

jye326 committed:
refactor: Jdbc 레포지토리 코드 제거 및 Jpa 레포지토리로 통합

- Jdbc 기반의 Reservation, ReservationTime, Theme, User 레포지토리 코드를 삭제하고 기존 Jpa 기반 레포지토리로 통합.
- 관련 테스트 코드 및 인프라 패키지 구조에서 불필요한 "Jpa" 네이밍 제거로 간결화 작업 수행.

jye326 committed:
fix(resources) : 내 예약 페이지 url 변경

 - /reservation-mine -> /reservation/mine

jye326 committed:
fix(resources) : 사용자 더미 데이터 추가

jye326 committed:
feat(resources) : 예약 대기 기능 클라이언트 코드 반영

jye326 committed:
docs(README) : 예약 대기 기능 문서 작성

jye326 committed:
feat(entity): Waiting 추가 및 Reservation 간 관계 매핑 추가

- Waiting 엔티티 생성 및 Reservation과 ManyToOne 관계 설정
- Reservation에 Waiting 리스트 추가 및 OneToMany 매핑 구현
- 대기 기능 관련 로직 및 설명 README에 업데이트

jye326 committed:
docs(README) : 예약 대기 기능 문서 작성 업데이트

jye326 committed:
refactor: Waiting 엔티티 제거 및 Reservation 내 waitingList 속성 삭제

jye326 committed:
feat: 예약 상태(Status) 관리 기능 추가

- 기존 ReservationStatus 제거 및 신규 Status enum 구현 (RESERVED, WAITING)
- 예약 생성 시 상태값(Status) 지정 기능 추가
- 예약 데이터와 관련된 엔티티, DTO, API 요청/응답 변경 및 통합 테스트 업데이트
- LocalDataInitializer, JavaScript 코드에서 상태값 반영 및 업데이트

예약 상태를 명확히 관리하기 위한 구조 개선과 상태값 기반 로직 구현 추가

jye326 committed:
feat: 예약 응답에 대기 인원 수 추가 및 불필요 주석 제거

jye326 committed:
feat(reservation): 예약 대기 기능 추가

- 본인 예약 조회 시 aheadCount 추가하여 대기 인원 조회 가능
- ReservationWithAhead DTO 추가 및 관련 로직 수정
- Repository 및 Service 계층에 대기 인원 계산 로직 반영
- 예약 목록 UI에서 대기 상태 시 대기 인원 표시

jye326 committed:
fix: 예약 관련 기능 수정

- 예약 대기 취소 API 호출 시 endpoint URL 설정 변경
- LocalDataInitializer에서 예약 상태를 WAITING에서 RESERVED로 변경
- 예약 대기 순서 계산 로직 수정 (COUNT 결과 타입 변경)

jye326 committed:
test: Reservation 생성 기능 수정 및 delete 메서드 단순화

- `Reservation.create` 메서드에 `Status`와 `LocalDateTime` 필드 추가
- 테스트 코드 수정: `Reservation.create` 호출 시 새로운 파라미터 반영
- `ReservationService.delete` 메서드에서 사용자 검증 로직 제거 및 단순화

jye326 committed:
feat(resources):예약 대기 관리 페이지 추가 및 관련 JS 로직 구현

jye326 committed:
feat: 관리자 대기 예약 관리 기능 추가

- 관리자 전용 대기 예약 조회 및 UI 페이지 추가
- 예약 취소 시 대기 예약 자동 갱신 로직 구현
- `ReservationService`, Repository, API에 관련 메서드 및 쿼리 추가
- 테스트 데이터에 대기 예약 상태 추가

jye326 committed:
refactor: import 구문 정리 및 코드 포매팅 개선

jye326 committed:
docs: README 포맷 수정 및 Reservation/Waiting 구조 정리 업데이트

jye326 committed:
merge step1

jye326 committed:
docs: README.md의 Waiting과 Reservation 관계 설명 보완

jye326 committed:
docs: README.md의 Waiting과 Reservation 관계 설명 보완

jye326 committed:
Merge branch 'step2' of https://github.com/jye326/spring-roomescape-waiting into step2

jye326 committed:
fix: 대기 예약 관련 서비스 및 메서드 리팩토링 (WaitingService 분리 및 비동기 처리 추가)

jye326 committed:
feat(resources) : 예약 삭제 후 목록 새로고침 확인 기능 추가

jye326 committed:
refactor: Reservation 관련 DTO 및 메서드 네이밍 변경 및 Status 사용 추가

- ReservationWithAhead → ReservationWithAheadDto로 이름 변경
- findAllWithFilter → findAllReservationWithFilter로 메서드 이름 변경
- Reservation 관련 insert에 Status 파라미터 추가 및 필터 메서드에 Status 조건 추가
- 관련 사용 코드 및 테스트 코드 전반 수정

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationDao.java:
```diff
@@ -22,17 +28,60 @@ public interface JpaReservationDao extends JpaRepository<Reservation, Id> {
+            @Param("date") ReservationDate date,
+            @Param("time") ReservationTime time,
+            @Param("theme") Theme theme
+    );
```
가장 오래전에 예약대기를 걸어 놓은 예약에 대해 예약상태로 업데이트를 해주는 쿼리군요.
이 로직을 Service 단에서 Reservation 객체를 불러와 상태를 변경한 뒤, 해당 객체를 저장하지 않고 바로 쿼리로 처리하신 특별한 이유가 있을까요?
이런 비즈니스 로직을 `SQL 쿼리로 직접 처리하는 방법`과 `Service 단에서 객체를 활용해 구현하는 방법`에는 어떤 차이가 있을까요?
또 각각의 접근 방식이 가지는 장단점은 무엇이라고 생각하시나요?

jye326 replied:
조회와 상태 변경 로직을 한번에 처리한 것은 DB 왕복 오버헤드를 줄이기 위함이었습니다. 저는 DB IO가 성능에 가장 크게 영향을 미치는 요소 중 하나라고 생각합니다. 따라서 DB에서 쿼리로 처리할 수 있는 간단한 로직은 쿼리로 풀어야 한다고 생각합니다. 반대로 쿼리를 사용하지 않고 Service 단에서 객체를 활용해 구현한다면 테스트도 용이하고, 특정 SQL의 의존도를 낮출 수 있어 다른 DB로의 마이그레이션이 수월하다는 이점이 있습니다.

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationDao.java:
```diff
@@ -22,17 +28,60 @@ public interface JpaReservationDao extends JpaRepository<Reservation, Id> {
+                FROM Reservation r
+                WHERE r.user.id = :userId
+            """)
+    List<ReservationWithAheadDto> findReservationsWithAhead(@Param("userId") Id userId);
```
presentation 계층의 dto가 infrastructure layer까지 와서 쓰이고 있네요. 이유가 있을까요?

jye326 replied:
패키지가 잘못 들어갔네요! 는 ReservationWithAheadDto 비즈니스 계층에서 사용되는 Dto입니다. 실제로 컨트롤러에서 presentation 계층 Dto인 ReservationMineResponse로 변환됩니다.

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationRepository.java:
```diff
@@ -32,8 +36,13 @@ public List<Reservation> findAll() {
-        return dao.findAllWithFilter(themeId, memberId, dateFrom, dateTo);
+    public List<Reservation> findAllReservationWithFilter(Id themeId, Id memberId, LocalDate dateFrom, LocalDate dateTo) {
+        return dao.findAllWithFilter(themeId, memberId, dateFrom, dateTo, Status.RESERVED);
+    }
```
```suggestion
    public List<Reservation> findAllReservationWithFilter(Id themeId, Id memberId, LocalDate dateFrom, LocalDate dateTo, Status status) {
        return dao.findAllWithFilter(themeId, memberId, dateFrom, dateTo, status);
    }
```
이렇게 status까지 parameter로 입력을 받으면, `findAllWaitingReservations`메서드와의 중복도 줄이고 하나의 메서드를 재사용 할 수 있을 것 같아요.

jye326 replied:
그렇네요! 중복도 방지할 수 있고 무엇보다 status 파라미터를 적절하게 넣어주는 것은 리포지토리보다 서비스의 책임이네요! 수정하였습니다. 감사합니다~

soominsohn commented on README.md:
```diff
@@ -0,0 +1,28 @@
+- [x] Reservation에 필드 status, createdAt을 추가
+    - [x] Reservation에 status 요청 필드 추가하기
+- [x] 대기 관련 로직을 처리하는 WaitingService를 추가하자. 
+
```
> LMS의 구현 힌트를 보고 처음에는 예약 대기를 Waiting 클래스를 만들어 관리하도록 구현하다가, 고민 끝에 Reservation 객체의 상태값(ENUM)으로 관리하도록 구현했습니다. 방탈출예약관리라는 지금의 도메인에 대해서 Waiting 클래스 분리는 필요 없다고 생각합니다. README에 관련한 고민을 의식의 흐름 기법으로 적어 놓았습니다. 관련해서 의견을 듣고 싶습니다!

저라면 Waiting과 Reservation을 분리했을 것 같아요.
예약 대기 내역(Waiting)과 예약 내역(Reservation)은 도메인은 물론, 데이터베이스 테이블 차원에서도 서로 다른 관심사를 가진 별개의 개념이라고 생각이 됩니다. Waiting은 단순히 예약의 상태 값이 아니라, 예약 확정을 기다리는 별도의 개념이며, 그에 따른 다양한 비즈니스 규칙을 가질 수 있어요.
지금은 필드 두 개만 추가하며 Reservation 객체로 간단히 해결한 것처럼 보일 수 있는데요, 실제로 예약 대기라는 개념이 확장되거나 복잡해지면, 단순 상태 정보 이상의 다양한 데이터와 로직(대기 취소 정책 등)이 추가될 수밖에 없습니다. 그렇게 된다면 이 로직들은 예약(Reservation)의 관심사가 아니겠죠?

비즈니스 관점 뿐만 아니라 지금처럼 Reservation 테이블 하나만 사용할 경우 예약 조회, 예약대기 조회, 예약대기 순번 조회, 수정, 삭제 등 모두 한 테이블에 부하가 가는데요, 이런 경우 성능 저하 이슈가 생길 수 있습니다. 

또, 만약 Waiting을 별도로 분리한다면, 저는 Reservation 객체를 직접 참조하지 않고 예약 날짜, 시간, 테마 정보를 별도로 저장하는 형태가 될 것 같습니다. 즉, Waiting은 이미 생성된 Reservation에 대한 대기가 아니라, 예약 확정 전의 날짜·시간·테마에 대한 대기 상태로 이해가 됩니다. 
띠용의 의견은 어떠신가요? 분리하는게 좋을 것 같나요? 아니면 이 구조 그대로 가는게 좋을 것 같나요?

jye326 replied:
저는 이 구조 그대로 분리할 수 없다고 생각합니다.

> 예약 대기 내역(Waiting)과 예약 내역(Reservation)은 도메인은 물론, 데이터베이스 테이블 차원에서도 서로 다른 관심사를 가진 별개의 개념이라고 생각이 됩니다.
> 

Reservation과 Waiting은 별도의 개념이 될 수 없다고 생각해요. 개념적으로 ‘어떤 예약에 대한 대기인지’가 반드시 포함되어있어야 하거든요. 두 개념을 분리할 시에 예약 날짜, 시간, 테마 정보가 중복되는 것도 이 때문이라고 생각합니다. 이것이 Reservation과 별도의 Entity로 관리되고 있는 ReservationTime과의 결정적 차이입니다. ReservationTime은 Reservation과 무관하게 생성되고 관리될 수 있는 엔티티입니다. Waiting은 다릅니다. 생성의 경우 반드시 Reservation이 선행되어야 하고, 중복된 Reservation 방지를 위해 Waiting이 Reservation으로 승인되는 과정도 Reservation 테이블에 종속적입니다. 오히려 현재 상태로 관리되고 있는 User와 Admin의 관계와 유사해 보입니다.

이러한 이유로 저는 두 개념을 분리하는 것이 부자연스러운 행위이며 분리가 가능은 하지만 지양해야 한다고 생각합니다. 

> 실제로 예약 대기라는 개념이 확장되거나 복잡해지면, 단순 상태 정보 이상의 다양한 데이터와 로직(대기 취소 정책 등)이 추가될 수밖에 없습니다.
> 

대기는 예약의 상태일 뿐이라는 입장은 변하지 않았지만, 그럼에도 대기만이 가지는 비즈니스 로직은 존재할 수 있습니다. 그러한 요구사항(ex. 예약 대기 시간 24시간 초과 시 자동 취소)은 얼마든지 만들 수 있죠! 하지만 현재로써는 그러한 독립적인 책임이 많을 것 같지 않고, 이를 위해 현재 동일한 것이라고 생각하는 도메인을 분리하는 것은 과도한 확장성 고려이고, 그보다는 예약을 처리하는 기존의 코드로 대기를 함께 관리할 수 있는 상태 구조의 이점이 더 크다고 생각합니다. 따라서 지금의 구조를 유지하되, 만약 대기만의 요구 사항이 추가되면 서비스 단에서 상태를 확인해 분기 처리할 것 같습니다. 현재는 WaitingService를 통해 이러한 서비스 단에서의 책임 분리를 시도하였습니다. 

> 비즈니스 관점 뿐만 아니라 지금처럼 Reservation 테이블 하나만 사용할 경우 예약 조회, 예약 대기 조회, 예약 대기 순번 조회, 수정, 삭제 등 모두 한 테이블에 부하가 가는데요, 이런 경우 성능 저하 이슈가 생길 수 있습니다.
> 

DB의 부담을 고려한다면 테이블만 분리하는 것을 고려할 것 같습니다. 기본적으로 Wainting은 ‘대기 중인 예약’이라는 입장은 유지하되, 필요하다면 테이블 부담을 줄이기 위해 Status를 기준으로 데이터만 테이블을 나누어 관리해 줄 것 같습니다.

> Waiting은 이미 생성된 Reservation에 대한 대기가 아니라, 예약 확정 전의 날짜·시간·테마에 대한 대기 상태
> 

이 구조도 괜찮다고 생각합니다. 저도 만약 분리를 하게 된다면 예약과 대기가 나란히 스펙(날짜·시간·테마)이라는 VO를 바라보는 구조를 선택할 것 같습니다. 그러나 분리 시 예약과 대기라는 서로 다른 도메인에서 스펙값이 불가피하게 중복되는 것이 찜찜할 것 같습니다.

soominsohn replied:
네, 말씀해주신 것처럼 현재 프로그램에서 Reservation과 Waiting을 분리하는 것이 어색하게 느껴진다면, 굳이 분리할 필요는 없다고 생각합니다.
지금 구조를 유지하면서, 서비스 레이어에서 상태를 기준으로 적절히 분기 처리하는 방향도 충분히 괜찮아 보이네요.
깊은 고민 해주어서 고마워요 👍

soominsohn commented on src/main/java/roomescape/business/model/vo/Status.java:
```diff
@@ -0,0 +1,6 @@
+public enum Status {
+    RESERVED,
+    WAITING
+}
```
만일 Reservation의 상태를 관리한다고 하면, `ReservationStatus`라는 이름이 더 적절할 것 같아요.
또, 다른 RESERVED, WAITING 외에 실패나 취소 등의 상태도 관리하면 좋을 것 같습니다.

jye326 replied:
말씀하신 대로 ReservationStatus 변경하는 것이 구체적이므로 변경했습니다! 또한 추후에 관리가 필요한 상태가 추가되면 해당 클래스에 추가할 계획입니다.

soominsohn commented on src/main/java/roomescape/business/service/WaitingService.java:
```diff
@@ -0,0 +1,31 @@
+
+    private final ReservationRepository reservationRepository;
+
+    @Async
```
`@Async`어노테이션은 어떻게 동작할까요?
우리 프로그램에서 필요한건가요?

jye326 replied:
저는 대기 승인이 자동적으로 이루어져야 한다고 생각하여 예약이 삭제되면 해당 예약의 대기 1번이 자동으로 예약으로 업데이트 되도록 설계했습니다. 그러나 예약 삭제는 대기의 업데이트와는 무관하게 이루어져야 하는 것이므로 삭제 후 비동기적으로 대기 업데이트를 호출할 수 있도록 `@Async` 어노테이션을 활용했습니다.

soominsohn replied:
네, 맞습니다. 띠용은 예약 삭제와 대기 예약 처리가 반드시 같은 트랜잭션 내에서 이루어질 필요는 없다고 생각하신 거죠.
예약을 취소하려고 하는데, 다른 사람의 예약 대기 삭제 과정에서 문제가 발생해 내 예약 취소가 실패한다면 사용자 입장에서 어색할 수 있겠네요! 👍 

하지만 이렇게 비동기 처리를 도입할 때는, 실패에 대한 대응과 예외 처리에 대해서도 고려해야 합니다. 
만약 예약 대기 업데이트가 실패하면, 해당 시간대의 예약들은 영원히 대기 상태로 남게 될 가능성이 큽니다.
이 경우 프로그램 차원에서 재처리 로직을 구현하거나, 어드민이 알아차리고 수동으로 대응할 수 있도록 실패를 감지할 수 있는 처리가 마련되어 있어야 해요.

비동기 처리는 동기 처리보다 훨씬 복잡도가 높기 때문에 학습하는데도 시간이 많이 걸릴거에요. 아주 좋은 생각이었지만, 이 학습 단계에서는 다른 부분에 대해 먼저 공부하는 것을 추천해요. 😄

jye326 replied:
말씀하신 대로 관련 로직을 전문적으로 담당하는 ‘비동기 처리 도구’가 존재함을 배웠습니다. 이런 툴을 현재 도입하기는 어렵기 때문에 에러 발생 시 서버에 로깅하기, update만 별도로 시행하는 api 생성 등을 고려할 수 있을 것 같습니다.

soominsohn commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -12,63 +20,66 @@
     }
 
-    public void delete(final String reservationId, final String userIdValue) {
+    public void deleteAndUpdateWaiting(final String reservationId) {
```
띠용은 `예약 대기 승인`을 자동으로 승인되게 구현해주신 것 같아요. 
수동 승인과 자동 승인이 있었을텐데 어떤 부분을 고려하셨는지 알려주실 수 있으실까요?

jye326 replied:
대기→ 예약의 변화는 결국 대기 순서대로 이루어지는 것이기 때문에 불필요한 관리자의 수동 승인 작업을 없애기 위해 자동으로 설계하였습니다.

soominsohn commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -12,63 +20,66 @@
                 .orElseThrow(() -> new NotFoundException(RESERVATION_NOT_EXIST));
-        if (!reservation.isSameReserver(userIdValue)) {
-            throw new AuthorizationException(AUTHORITY_LACK);
-        }
-        reservationRepository.deleteById(id);
+        delete(id);
+        waitingService.updateWaitingReservations(reservation);
+    }
```
> `OSIV, 트랜잭션, 프록시 객체 등`의 개념을 체감할 수 있는 부분을 찾지 못했습니다.

여기서는 `트래잭션`에 대해 공부해보죠!
자. 이 코드에서
1. 기존에 존재하는 Reservation을 삭제하고
2. 예약대기 상태인 Reservation을 예약상태로 변경하게 되는데
1번은 성공하고 2번은 실패하면 어떻게 될까요? 


트랜잭션에 대해 학습해보고, 스프링이 지원하는 `@Transacitonal` 어노테이션의 사용법에 대해서도 같이 학습해봅시다.

jye326 replied:
예약 삭제는 대기 업데이트와 독립적으로 실행되어야 한다는 점을 고려하여 `updateWaitingReservations` 에 `@Async` 를 걸어 주었습니다. 따라서 2번의 실패는 1번에 영향을 주지 않습니다.

트랜잭션과 `@Transcational`에 관하여 학습해 보았습니다! 트랜잭션은 데이터베이스의 일관성을 보장하기 위한 논리적 작업 단위입니다. `@Transactional`은 Spring이 메서드를 하나의 트랜잭션으로 관리하기 위한 어노테이션이며 선언적으로 AOP를 구현합니다.

soominsohn replied:
트랜잭션과 `@Transactional`에 대해 찾아보셨군요!
`@Transactional`에는 트랜잭션 전파(propagation) 옵션이나 격리 수준(isolation level) 같은 세부 설정도 있는데요, 
트랜잭션을 어떤 단위(예: 클래스, 메서드 등)에 적용하고, 어떤 전파 옵션, 격리 수준을 사용하는지 찾아보고 비교해보는 것도 큰 도움이 될 것 같아요. 이번 미션의 요구사항은 아닌 것 같기 때문에 천천히 학습해보시면 좋을 것 같습니다.

jye326 replied:
아직 각각에 대해 디테일하게 이해하지는 못했지만 기초적인 개념을 공부해 보았습니다.

전파 옵션은 기존 트랜잭션 존재 유무에 따른 트랜잭션의 생성 여부를 정할 때 사용되는 옵션입니다.
격리 수준은 서로 다른 트랜잭션이 동시에 실행될 때 서로의 변경 내용을 얼마나 보여줄 것인가를 설정하는 옵션입니다!

soominsohn commented on src/main/java/roomescape/business/model/entity/Reservation.java:
```diff
@@ -1,11 +1,23 @@
+import jakarta.persistence.Entity;
+import jakarta.persistence.EnumType;
+import jakarta.persistence.Enumerated;
+import jakarta.persistence.ManyToOne;
```
> `OSIV, 트랜잭션, 프록시 객체` 등의 개념을 체감할 수 있는 부분을 찾지 못했습니다.

여기서는 `프록시 객체, 지연로딩`에 대해 학습해보아요.
띠용은 이번 미션을 진행하면서 이번 미션에서 ManyToOne 연관관계를 활용해 Reservation과 관련된 객체를 조회하셨는데요,
JPA는 테이블에 저장된 외래 키(id) 값을 어떻게 실제 객체로 매핑할까요?
또한, 객체를 가져올 때 사용하는 FetchType 옵션에는 어떤 종류가 있으며, 각각 어떻게 동작하는지도 학습해보아요.

jye326 replied:
JPA구현체(hibernate)가 외래키 ↔ 객체 매핑을 처리할 때 프록시 객체와 지연로딩을 사용합니다. 이는 엔티티에 연관되어 있지만 당장에 필요하지 않은 필드 엔티티를 불필요하게 DB 요청하지 않기 위함입니다. 예를 들어 Member{name, email}라는 엔티티가 있고, Member Id만 조회할 때는 굳이 name과 email을 알 필요가 없습니다. 따라서 JPA는 온전한 Member엔티티가 아닌 name과 eamil의 외래키 값만 들고 있는 프록시 객체 `ex.Member$Proxy`를 불러옵니다. 추후 member.getName()과 같이 실제 내부 필드 값이 필요한 상황이 되었을 때 외래키 값으로 Join 하는 쿼리를 날려 객체를 완성합니다. 이것을 지연 로딩이라고 하며 `FetchType.lazy` 로 사용할 수 있습니다. 지연 로딩은 상황에 따라 N+1 문제 등이 발생할 수 있습니다.

반대로 FetchType.Eager은 한 번에 연관된 모든 엔티티를 조회하는 것입니다.

soominsohn commented on src/main/java/roomescape/business/service/ReservationService.java:
```diff
@@ -12,63 +20,66 @@
-        Reservation reservation = Reservation.create(user, date, reservationTime, theme);
+        Reservation reservation = Reservation.create(user, date, reservationTime, theme, status, LocalDateTime.now());
         reservationRepository.save(reservation);
         return ReservationDto.fromEntity(reservation);
```
> `OSIV, 트랜잭션, 프록시 객체` 등의 개념을 체감할 수 있는 부분을 찾지 못했습니다.

여기서는 `OSIV`에 대해 학습해보아요~
`ReservationDto.fromEntity(reservation);`처럼 엔티티를 DTO로 변환하지 않고, Reservation 엔티티를 그대로 컨트롤러에 넘겨준 후 뷰에서 필요한 데이터에 접근하려고 하면 어떤 문제가 발생할까요?

이 상황에서 JPA의 OSIV가 어떻게 동작하는지, 그리고 OSIV가 없을 때 발생하는 예외는 무엇일지 찾아봅시다~

jye326 replied:
`Reservation `엔티티를 그대로 컨트롤러/뷰에 넘겨준다면 지연 로딩이 적용된 프록시 객체가 서비스 레이어를 통과하고, 작업이 종료되어 영속성 컨텍스트를 관리하던 `EntityManager `= `Session `이 종료됩니다. 이후 뷰에서 필요한 데이터에 접근하려고 하면 연결이 끊겼기 때문에 `LazyInitializationException `이 발생합니다. `View `랜더링 시까지 세션을 유지하는 OSIV 옵션을 킨다면 이러한 문제를 방지할 수 있습니다.

soominsohn reviewed (changes requested):
띠용 안녕하세요~
이번 미션도 잘 구현해 주셨네요.
코멘트 몇 개 남겼으니 확인 부탁드려요. 😄

jye326 committed:
refactor: `ReservationWithAheadDto` 패키지 이동 및 예약 필터에 상태 조건 추가

- `ReservationWithAheadDto`를 `presentation`에서 `business` 패키지로 이동
- 예약 필터 (`findAllReservationWithFilter`)에 상태 (`Status`) 조건 추가
- 불필요한 `findAllWaitingReservations` 메서드 제거 및 관련 로직 통합
- 관련 테스트 및 서비스 로직 수정

jye326 committed:
fix: 예약 생성 시 대기 예약 상태 업데이트 로직 추가

- for 예약이 안되어 있는 상황에서 예약 대기 요청이 먼저 들어올 경우

jye326 committed:
refactor: `Status` 클래스명을 `ReservationStatus`로 변경

- `Status` 클래스명을 더 명확한 의미의 `ReservationStatus`로 리팩토링
- 관련 클래스, 메서드, 변수 및 테스트 코드 이름과 타입 변경
- JPQL 및 쿼리 파라미터에서도 해당 변경 반영

soominsohn commented on src/main/java/roomescape/infrastructure/JpaReservationDao.java:
```diff
@@ -22,17 +28,59 @@ public interface JpaReservationDao extends JpaRepository<Reservation, Id> {
 
     boolean existsByThemeId(Id themeId);
+
+    @Transactional
```
`@Transactional` 어노테이션이 이곳에만 붙어 있네요!
어떤 기준으로 해당 위치에 적용하셨나요?
이왕 사용하신 김에, `@Transactional`에 대해 더 학습해보고, 다른 곳에서도 사용될 수 있을지 찾아보아요!

jye326 replied:
작성 당시에는 단순히 update쿼리를 작성할 때는 `@Modifying` 과 `@Transactional` 이 필요하다고 하길래 일단 적용했습니다.

`@Transactional` 에 대해 알아보니, 기존에 `@Transactional` 이 없어도 기능이 동작했던 것은 SpringDataJpa가 `findBy…`, `findAll…`, `@Query` 등으로 정의한 SELECT 계열 조회 메서드와 `save()`, `saveAll()`, `deleteById()`, `deleteAll()` 과 같은 기본 삽입, 삭제 메서드에 대해 자동으로  `@Transactional` 을 달아주었기 때문이었습니다. 그러나 JPA가 지원하는 기본 네이밍이 아니며 직접 JPQL로 쿼리를 작성한 경우에는 메서드 이름으로 해당 메서드의 동작 과정을 파악할 수 없어, 기본적으로 `@Transactional(readOnly=ture)` 가 붙게 됩니다. 따라서 update 쿼리의 경우는 `@Modifying` 을 붙여 JPQL을 executeUpdate 모드로 변경하고, `@Transactional(readOnly=false)` 를 명시해야 합니다.

또한 일반적으로 SpringDataJpa의 기본 적용과는 별개로, 트랜잭션은 서비스 단위로 관리되기 때문에 리포지토리가 아니라 서비스 레이어에서 명시적으로 `@Transactional` 을 선언합니다. 또한 조회용 메서드에는 영속성 컨텍스트의 더티체킹이 필요하지 않기 때문에 `@Transactional(readOnly = true)` 를 선언해 최적화 합니다.

위 학습내용을 반영해 리팩토링했습니다!

soominsohn replied:
와우! 짧은 시간에 많은 부분을 학습해주셨네요! 👍 감동입니다!

soominsohn reviewed (changes requested):
안녕하세요 띠용~
코멘트 몇 개 더 남겼으니 확인부탁드려요~
다음 리뷰 요청엔 머지할 것 같습니다! 👍

jye326 committed:
refactor: 서비스 계층 전반에 @Transactional 추가 및 메서드에 readOnly 속성 설정

soominsohn reviewed (approved):
안녕하세요 띠용~
이번 미션 수고 많으셨습니다!
띠용이 정리해주신 내용보면서 저도 많이 배웠습니다~
앞으로도 파이팅하세요!
